secure_message/resources/threads.py

```diff
@@ -21,13 +22,77 @@ def get(thread_id):
+        if conversation_metadata is None:
+            abort(404)
 
         logger.info("Successfully retrieved messages from thread", thread_id=thread_id, user_uuid=g.user.user_uuid)
         messages = []
         for message in conversation.all():
             msg = message.serialize(g.user, body_summary=False)
             messages.append(msg)
-        return jsonify({"messages": add_users_and_business_details(messages)})
+
+        if conversation_metadata.is_closed:
+            return jsonify({"messages": add_users_and_business_details(messages),
+                            "is_closed": conversation_metadata.is_closed,
+                            "closed_by": conversation_metadata.closed_by,
+                            "closed_by_uuid": conversation_metadata.closed_by_uuid,
+                            "closed_at": conversation_metadata.closed_at})
+
+        return jsonify({"messages": add_users_and_business_details(messages),
+                        "is_closed": False})
+
+    @staticmethod
+    def patch(thread_id):
+        """Modify conversation metadata"""
+        bound_logger = logger.bind(thread_id=thread_id, user_uuid=g.user.user_uuid)
+        bound_logger.info("Validating request")
+        if not g.user.is_internal:
+            bound_logger.info("Thread modification is forbidden")
+            abort(403)
+        if request.headers['Content-Type'].lower() != 'application/json':
+            bound_logger.info('Request must set accept content type "application/json" in header.')
+            raise BadRequest(description='Request must set accept content type "application/json" in header.')
+
+        bound_logger.info("Retrieving metadata for thread")
+        request_data = request.get_json()
+        metadata = Retriever.retrieve_conversation_metadata(thread_id)
+        if metadata is None:
+            abort(404)
+        ThreadById._validate_request(request_data, metadata)
+
+        bound_logger.info("Attempting to modify metadata for thread")
+        if request_data.get('is_closed'):
+            bound_logger.info("About to close conversation")
+            Modifier.close_conversation(metadata, g.user)
+        else:
+            bound_logger.info("About to re-open conversation")
+            Modifier.open_conversation(metadata, g.user)
+
+        bound_logger.info("Thread metadata update successful")
+        bound_logger.unbind('thread_id', 'user_uuid')
+        return '', 204
+
+    @staticmethod
+    def _validate_request(request_data, metadata):
+        """Used to validate data within request body"""
+        bound_logger = logger.bind(thread_id=metadata.id, user_uuid=g.user.user_uuid)
+        # Check if it's empty
+        if not request_data:
+            bound_logger.error('No properties provided')
+            raise BadRequest(description="No properties provided")
+        if not isinstance(request_data['is_closed'], bool):
```

Comment:
If you send a request without the `is_closed` property e.g. `{ "fake_property": 12 }` a 500 error is generated. It would be better to catch the `KeyError` when accessing `request_data['is_closed']` and return a 400 error.

Reply:
Fixed it